### PR TITLE
Topic/leave dots alone

### PR DESCRIPTION
**Pull request by EduardSergeev**
Remove "special treatment" dots in S3 Key. Pass Key as is.

### CODE REVIEW

File: core/src/Network/AWS/Data/Internal/URI.hs [hunk 1]
```diff
@@ -38,19 +36,6 @@ collapsePath bs
          | BS.last bs == sep = x `BS.snoc` sep
          | otherwise         = x
 
-    go acc c | c == dot  = acc
```
Comment (EduardSergeev) [inline core/src/Network/AWS/Data/Internal/URI.hs]: Removed "dot processing".

**Comment by charleso** [comment]
As mentioned IRL I would wait for confirmation from Brendan before merging

**Comment by charleso** [comment]
For reference, when it was introduced:

https://github.com/ambiata/amazonka/commit/8ad6b4c919263818f6c4486453e700d169d57f37

**Comment by EduardSergeev** [comment]
@charleso 
Actually it looks like the behaviour is different (the correct one) in `develop`: https://github.com/brendanhay/amazonka/blob/develop/core/src/Network/AWS/Sign/V4.hs#L249-L253
where `collapsePath` is: https://github.com/brendanhay/amazonka/blob/develop/core/src/Network/AWS/Data/Path.hs#L78-L93
I haven't tested it but it looks like it won't "collapse" paths for S3.

**Comment by EduardSergeev** [comment]
Bump @charleso, @nhibberd, @brendanhay Do you think it is OK to merge this change to the current version, just to move mismi to this old version of `amazonka` as a first step before upgrading to `develop` version as a next step? Our tests won't pass without this fix.

**Comment by charleso** [comment]
As discussed IRL I'm OK with this being merged, although it would be good to confirm that it's the _exact_ same behaviour as `1.0/develop`. Sounds like we might want to avoid this step for S3 calls only and possibly confirm with Brendan as well.

**Comment by EduardSergeev** [comment]
@charleso Good point.
I've copied logic from `develop` so we don't preprocess paths for S3.

**Comment by brendanhay** [comment]
Let me know if I can do anything to help. The reason for now not collapsing S3 paths is due to how the other SDKs handle it:

* [botocore/utils.py](https://github.com/boto/botocore/blob/7ae26d6c09102847f5e61eefc94a2822faa3354f/botocore/utils.py#L50-L94)
* [botocore/auth.py](https://github.com/boto/botocore/blob/7ae26d6c09102847f5e61eefc94a2822faa3354f/botocore/auth.py#L369)

**Comment by EduardSergeev** [comment]
@brendanhay Could you please clarify the situation with S3 keys in `develop` (soon to be 1.0.1?). `develop` version makes an exception for S3 paths passing them "as is", is that correct? While it does precess `.` and `..` for other services. Am I reading the code correctly?

**Comment by brendanhay** [comment]
Yes, that is correct. The signing calculations now behave identically to `botocore` outlined in the two `.py` files I linked above.

To create the canonical path as part of canonical request construction:

* All services have the path URL encoded and escaped for segments between the `/` delimiter.
*  All services (except S3) have their paths collapsed.

I cannot actually offer up a consistent reason why `V4` signing has the S3 exception, only that it exists in all the SDKs.

**Pull request merged**